File: solutions/Leetcode_97/leetcode_97.cc
#include <string>
#include <vector>

using namespace std;


class Solution1
{
public:
    bool isInterleave(string s1, string s2, string s3)
    {
        if (s1.length() + s2.length() != s3.length())
            return false;
        vector<vector<bool>> invalid(s1.length()+1, vector<bool>(s2.length()+1, false));
        return compare(s1, s2, s3, 0, 0, 0, invalid);
    }

private:
    bool compare(string &s1, string &s2, string &s3, int i, int j, int k, vector<vector<bool>> &invalid)
    {
        if (invalid[i][j])
            return false;
        if (k == s3.length())
            return true;
        bool is_valid = i < s1.length() && s1[i] == s3[k] && compare(s1, s2, s3, i+1, j, k+1, invalid) ||
            j < s2.length() && s2[j] == s3[k] && compare(s1, s2, s3, i, j+1, k+1, invalid);
        if (!is_valid)
            invalid[i][j] = true;
        return is_valid;
    }
};


class Solution2
{
public:
    bool isInterleave(string s1, string s2, string s3)
    {
        int s1_len = s1.size(), s2_len = s2.size(), s3_len = s3.size();
        if (s1_len + s2_len != s3_len)
            return false;
        bool invalid[s2_len+1];
        invalid[0] = true;
        for (int i = 1; i < s2_len+1; i++)
            invalid[i] = invalid[i-1] && s2[i-1] == s3[i-1];
        for (int i = 1; i < s1_len+1; i++)
        {
            invalid[0] = invalid[0] && s1[i-1] == s3[i-1];
            for (int j = 1; j < s2_len+1; j++)
                invalid[j] = (invalid[j-1] && s2[j-1] == s3[i+j-1]) || (invalid[j] && s1[i-1] == s3[i+j-1]);
        }
        return invalid[s2_len];
    }
};

File: solutions/Leetcode_97/leetcode_97.py
class Solution1:
    def is_interleave(self, s1: str, s2: str, s3: str):
        s1_len = len(s1)
        s2_len = len(s2)
        s3_len = len(s3)
        if s1_len + s2_len != s3_len:
            return False
        invalid = [[False] * (s2_len+1) for _ in range(s1_len+1)]
        def compare(i, j, k):
            if invalid[i][j]:
                return False
            if k == s3_len:
                return True
            is_valid = (i < s1_len and s1[i] == s3[k] and compare(i+1, j, k+1)
                        or j < s2_len and s2[j] == s3[k] and compare(i, j+1, k+1))
            if not is_valid:
                invalid[i][j] = True
            return is_valid
        return compare(0, 0, 0)


class Solution2:
    def is_interleave(self, s1, s2, s3):
        s1_len = len(s1)
        s2_len = len(s2)
        s3_len = len(s3)
        if s1_len + s2_len != s3_len:
            return False
        dp = [True] * (s2_len+1)
        for j in range(1, s2_len+1):
            dp[j] = dp[j-1] and s2[j-1] == s3[j-1]
        for i in range(1, s1_len + 1):
            dp[0] = dp[0] and s1[i-1] == s3[i-1]
            for j in range(1, s2_len + 1):
                dp[j] = (dp[j] and s1[i-1] == s3[i+j-1]) or (dp[j-1] and s2[j-1] == s3[i+j-1])
        return dp[-1]
                

